File: luaCollectors/cpuload.lua
cfg = {}
cfg["name"] = "cpuload"
cfg["poolfreq"] = 10
cfg["history"]  = 150


function declare ()
	this.addRessource("avg", "Load average", "loadaverage")
	this.addProperty( "avg", "avg1",  "Load average (1mn)", "number")
	this.addProperty( "avg", "avg5",  "Load average (5mn)", "number")
	this.addProperty( "avg", "avg15", "Load average (15mn)", "number")
end


function collect ()
	local f = assert(io.open("/proc/loadavg", "r"))
	local t = f:read("*all")
	f:close()
	a,b,c = t:match("(%d+.%d+) (%d+.%d+) (%d+.%d+)")
	this.nextValue("avg")
	this.setDProperty("avg", "avg1", a)
	this.setDProperty("avg", "avg5", b)
	this.setDProperty("avg", "avg15", c)
end

File: collectors/cpuload.cpp
#include "collectors.h"
#include <fstream>
#include <iostream>
#include <chrono>

using namespace watcheD;
namespace collectors {

class CpuLoadCollector : public Collector {
public:

	CpuLoadCollector(std::shared_ptr<HttpServer> p_srv, Json::Value* p_cfg) : Collector("cpuload", p_srv, p_cfg, 150, 10) {
		addRessource("avg", "Load average", "loadaverage");
		ressources["avg"]->addProperty("avg1",  "Load average (1mn)", "number");
		ressources["avg"]->addProperty("avg5",  "Load average (5mn)", "number");
		ressources["avg"]->addProperty("avg15", "Load average (15mn)", "number");
		addGetMetricRoute();
	}

	void collect() {
		std::string	line;
		std::ifstream	infile("/proc/loadavg");
		while(infile.good() && getline(infile, line)) {
			ressources["avg"]->nextValue();
			int pos = line.find(" ",0);
			ressources["avg"]->setProperty("avg1", stod(line.substr(0, pos)));
			ressources["avg"]->setProperty("avg5", stod(line.substr(pos, line.find(" ",pos+1)-pos)));
			pos = line.find(" ",pos+1);
			ressources["avg"]->setProperty("avg15", stod(line.substr(pos, line.find(" ",pos+1)-pos)));
		}
		if (infile.good())
			infile.close();
	}
};

MAKE_PLUGIN_COLLECTOR(CpuLoadCollector, cpuload)

}
